docs/contributing/code-quality-principles.md

```diff
@@ -0,0 +1,167 @@
+
+You can find an old code that doesn't follow these principles and we are ok with that as we are not able to refactor the whole framework at once.
+It may also happen that changed code cannot follow a principle because it has to be compatible with our [backward compatibility promise](backward-compatibility-promise.md),
+backward compatibility is more important than the code quality.
+
+These principles are not rules, they can be violated when it has a reason.
+The reason has to be discussed with the reviewer and may be written in the commit message.
+
+## Principles
+
+* [Dependency injection](#dependency-injection)
+* [Strict types](#strict-types)
+* [Don't repeat yourself](#dont-repeat-yourself)
+* [Dependency inversion](#dependency-inversion)
+
+## Dependency injection
+
+Dependency injection (or DI) means that an object doesn't create dependencies by itself but receives dependencies by a provider.
+We use constructor DI only.
+
+```diff
+class LocalizedType
+{
+    /**
+     * @var \Shopsys\FrameworkBundle\Model\Localization\Localization
+     */
+    private $localization;
+
+-   public function __construct()
+-   {
+-       $this->localization = new Localization();
+-   }
++   /**
++   * @param \Shopsys\FrameworkBundle\Model\Localization\Localization $localization
++   */
++   public function __construct(Localization $localization)
++   {
++       $this->localization = $localization;
++   }
+
+    // ...
+}
+```
+
+### Reasons
+
+It may be difficult to create the dependency - it may have also dependencies or configuration.
+Once creating the dependency is not the class responsibility, the class code is much more straight-forward.
+
+We inject only dependencies that the class needs, it is forbidden to pass the service locator to the class.
+With service locator injection it is difficult to understand what the class really do (or need) and such class is difficult to test.
+
+We use constructor DI so the class is always in a consistent state.
+
+It is essential to provide code extensibility in the framework and this is achieved with proper DI.
+
+### Exceptions
+
+We don't respect this principle in tests as it is impossible to pass dependencies to PHPUnit tests via the constructor.
+In tests we use `$this->getContainer()->get(...)`.
+
+## Strict types
+
+`strict_types` is a directive that enforces PHP to check the type of scalar types (`int`, `float`, `string`, `bool`) strictly.
+
+```diff
++ declare(strict_types=1);
+
+final class Money implements JsonSerializable
+{
+    // ...
+
+    public static function createFromFloat(float $float, int $scale): self
+    // ...
+}
+```
+
+If the caller of `createFromFloat()` method provides different arguments than declared, PHP throws a `TypeError`.
+This leads to safer code.
+
+### Reasons
+
+We use scalar types (`int`, `float`, `string`, `bool`) as much as possible in new code.
```

Comment:
```suggestion
We use scalar types (`int`, `float`, `string`, `bool`) as much as possible in a new code.
```

Reply:
Grammarly confirms both with and without the article `a` as valid... I think there shouldn't be the article because of the `new`